python/BioSimSpace/Sandpit/Exscientia/FreeEnergy/_restraint_search.py

```diff
@@ -0,0 +1,1113 @@
+                            boresch_dof_dict[pair][dof]["values"].std()
+                        # Exclude variance of internal angles as these are not restrained
+                        if (dof != "thetaR" and dof != "thetaL"):
+                            boresch_dof_dict[pair]["tot_var"] += \
+                            boresch_dof_dict[pair][dof]["sd"] ** 2
+                        # Assume Gaussian distributions and calculate force constants for harmonic potentials
+                        # so as to reproduce these distributions
+                        boresch_dof_dict[pair][dof]["k"] = 0.593 / (
+                                    boresch_dof_dict[pair][dof][
+                                        "sd"] ** 2)  # RT at 289 K is 0.593 kcal mol-1
+
+        ### Filter, Pick Optimum Degrees of Freedom, and Select Force Constants Based on Variance, Select Equilibrium Values
+
+        # Order pairs according to variance
+        pairs_ordered_boresch_var = []
+        for item in sorted(boresch_dof_dict.items(),
+                           key=lambda item: item[1]["tot_var"]):
+            pairs_ordered_boresch_var.append(item[0])
+            # print(f'Pair: {item[0]}, total variance: {boresch_dof_dict[item[0]]["tot_var"]}')
+
+        # Filter out r < 1, theta >150 or < 30
+        selected_pairs_boresch = []
+        for pair in pairs_ordered_boresch_var:
+            cond_dist = boresch_dof_dict[pair]["r"]["avg"] > 1
+            avg_angles = []
+            # angles = ["thetaA", "thetaB", "thetaR","thetaL"] # also check internal angles
+            angles = ["thetaA",
+                      "thetaB"]  # May also be good to check internal angles, although will be much stiffer
+            for angle in angles:
+                avg_angles.append(boresch_dof_dict[pair][angle]["avg"])
+            cond_angles = list(
+                map(lambda x: (x < 2.62 and x > 0.52), avg_angles))
+            if cond_dist and all(cond_angles):
+                selected_pairs_boresch.append(pair)
+
+        # Plotting
+        dof_to_plot = ["r", "thetaA", "thetaB", "phiA", "phiB",
+                       "phiC"]  # Can also plot thetaR and thetaL
+        n_dof = len(dof_to_plot)
+        pair = selected_pairs_boresch[0]
+
+        # Plot histograms
+        fig, axs = _plt.subplots(1, n_dof, figsize=(16, 4), dpi=500)
+        for i, dof in enumerate(dof_to_plot):
+            axs[i].hist(boresch_dof_dict[pair][dof]["values"], bins=10)
+            axs[i].axvline(x=boresch_dof_dict[pair][dof]["avg"], color='r',
+                           linestyle='dashed', linewidth=2, label="mean")
+            if dof == "r":
+                axs[i].set_xlabel("r ($\AA$)")
+            else:
+                axs[i].set_xlabel(f"{dof} (rad)")
+            axs[i].set_ylabel("Num Vals")
+            axs[i].legend()
+        fig.tight_layout()
+        fig.savefig(f'{work_dir}/boresch_dof_hist.png', facecolor="white")
```

Comment:
Does plotting always work? What happens if the user is logged in on a server without an X session?

Reply:
Not addressed as it is Finlay's code.